ocaml/forkexecd/helper/vfork_helper.c

```diff
@@ -335,14 +335,44 @@ reset_signal_handlers(void)
 static void
 clear_cgroup(void)
 {
-    int fd = open("/sys/fs/cgroup/systemd/cgroup.procs", O_WRONLY|O_CLOEXEC);
-    if (fd >= 0) {
-        char string_pid[32];
-        int ignored __attribute__((unused));
-        sprintf(string_pid, "%d\n", (int) getpid());
-        ignored = write(fd, string_pid, strlen(string_pid));
+    // list of files to try, terminated by NULL
+    static const char *cgroup_files[] = {
+        "/sys/fs/cgroup/systemd/cgroup.procs",
+        "/sys/fs/cgroup/cgroup.procs",
+        NULL
+    };
+
+    char string_pid[32];
+    int last_error = 0;
+    const char *last_error_msg = NULL;
+    const char *last_fn = NULL;
+
+    snprintf(string_pid, sizeof(string_pid), "%d\n", (int) getpid());
+
+    for (const char **fn = cgroup_files; *fn != NULL; ++fn) {
+        last_fn = *fn;
+        int fd = open(*fn, O_WRONLY|O_CLOEXEC);
+        if (fd < 0) {
+            last_error = errno;
+            last_error_msg = "Error opening file %s";
+            continue;
+        }
+
+        ssize_t written = write(fd, string_pid, strlen(string_pid));
+        if (written < 0) {
+            last_error = errno;
+            last_error_msg = "Error writing file %s";
+        }
         close(fd);
```

Comment:
Errors can also be reported on `close` (EIO/ENOSPC/etc.). I don't know whether Linux would ever do that on sysfs, but in general we should probably check the return value of close and handle errno (we'll likely get a compiler warning about it otherwise).

Reply:
No, you don't get errors from the compiler for `close`. `close` is a bit odd in handling errors, usually functions try to not perform operations or try to leave the stare unaltered/stable on exit. `close` always closes the file descriptor. No matter if any error happens during the operation.
About the error, no, for sysfs is not worth checking.